From 4096d0192363bdc031ab2d13bed5b3a4b109f255 Mon Sep 17 00:00:00 2001
From: Ulf Hermann <ulf.hermann@digia.com>
Date: Wed, 17 Sep 2014 11:46:38 +0200
Subject: Add methods for finding out if scene graph context has been
 initialized

We cannot rely on the OpenGL context for that as there may be
non-OpenGL scene graphs. In order for those to announce their readiness
QSGRenderContext::isValid() has to be virtual.

Change-Id: Ica35d8590e3f06edb54d57d93898d52e753f0407
Reviewed-by: Gunnar Sletta <gunnar@sletta.org>
---
 src/quick/items/qquickwindow.cpp    | 9 +++++++++
 src/quick/items/qquickwindow.h      | 1 +
 src/quick/scenegraph/qsgcontext_p.h | 2 +-
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/quick/items/qquickwindow.cpp b/src/quick/items/qquickwindow.cpp
index 6c6b34308..ca0a81f2f 100644
--- a/src/quick/items/qquickwindow.cpp
+++ b/src/quick/items/qquickwindow.cpp
@@ -2925,6 +2925,15 @@ QOpenGLContext *QQuickWindow::openglContext() const
     return d->context->openglContext();
 }
 
+/*!
+    Returns true if the scene graph has been initialized; otherwise returns false.
+ */
+bool QQuickWindow::isSceneGraphInitialized() const
+{
+    Q_D(const QQuickWindow);
+    return d->context != 0 && d->context->isValid();
+}
+
 /*!
     \fn void QQuickWindow::frameSwapped()
 
diff --git a/src/quick/items/qquickwindow.h b/src/quick/items/qquickwindow.h
index 108bcec58..82f2e29d9 100644
--- a/src/quick/items/qquickwindow.h
+++ b/src/quick/items/qquickwindow.h
@@ -139,6 +139,7 @@ public:
     bool isPersistentSceneGraph() const;
 
     QOpenGLContext *openglContext() const;
+    bool isSceneGraphInitialized() const;
 
     void scheduleRenderJob(QRunnable *job, RenderStage schedule);
 
diff --git a/src/quick/scenegraph/qsgcontext_p.h b/src/quick/scenegraph/qsgcontext_p.h
index 184b8248d..91788d82d 100644
--- a/src/quick/scenegraph/qsgcontext_p.h
+++ b/src/quick/scenegraph/qsgcontext_p.h
@@ -93,7 +93,7 @@ public:
 
     QOpenGLContext *openglContext() const { return m_gl; }
     QSGContext *sceneGraphContext() const { return m_sg; }
-    bool isValid() const { return m_gl; }
+    virtual bool isValid() const { return m_gl; }
 
     virtual void initialize(QOpenGLContext *context);
     virtual void invalidate();
-- 
cgit v1.2.1
